### PR TITLE
fix(API): return 500 response to server for uncaught exceptions

### PR DESCRIPTION
**Pull request by csojinb**
Per newly-understood WSGI spec, applications should trap internal errors instead of passing them on to the server. Add a default exception handler for the base Python ``Exception``, which is placed first in the error handler list (and is thus evaluated last, per LIFO operation).

The 500 error response sends no specific error information back to the client, but it does log a traceback to ``wsgi.errors``. This logging behavior currently has no tests.

Update error-handling documentation to reflect new default behavior. No "recipe" is included as yet -- need more details on requirements here.

Currently, errors are serialized in JSON or XML, depending on the Accept header in the request. It defaults to JSON if the request does not prefer XML, even if the app's default media type is XML. Changing this behavior appears to be non-trivial, so it is not attempted here, though it was discussed during development.

Addresses https://github.com/falconry/falcon/issues/1507

### CODE REVIEW

File: falcon/api.py [hunk 1]
```diff
@@ -186,6 +187,7 @@ def __init__(self, media_type=DEFAULT_MEDIA_TYPE,
         self.resp_options.default_media_type = media_type
 
         # NOTE(kgriffs): Add default error handlers
+        self.add_error_handler(Exception, self._python_error_handler)
```
Comment (jmvrbanac) [inline falcon/api.py]: I'm wondering if this should be `Exception` or `BaseException`... @kgriffs thoughts?

Comment (vytas7) [reply falcon/api.py]: I would suggest just _Exception_. The base one could make the WSGI app catch stuff like system exit, keyboard interrupt / Ctrl+C etc, and return JSON in those cases. That doesn't sound like safe default behaviour, but maybe we document how to, in case anyone still wants that thing.

Comment (csojinb) [reply falcon/api.py]: So, are we okay with this as-is?

File: falcon/api.py [hunk 2]
```diff
@@ -610,6 +613,11 @@ def set_error_serializer(self, serializer):
             property value, and it may choose to override the
             representation logic.
 
+        Note:
```
Comment (csojinb) [inline falcon/api.py]: https://github.com/falconry/falcon/blob/f078b2e6f1d54100f097b2fbb560a2b56fe7d922/falcon/api.py#L616-L619

This is not directly related to the changes in this PR, but I didn't see that this specific interaction was documented, so it seemed worth including.

**Comment by csojinb** [comment]
So, the hug tests are failing on this. Do I need to make a PR to there before this can be merged?

**Comment by csojinb** [comment]
Okay, so a brief scan of the test output suggests that a number of hug tests are depending on the behavior where falcon passes out I caught errors. 

I think that I could make a PR to hug that would override the now-default (in this PR) behavior to catch those exceptions for the purposes of their tests. (Hopefully I can do that with minimal disruption.) However, that does create the annoying situation where the default HTTPError and HTTPStatus handlers are also overridden, which is just kinda gross.

So, I’m thinking maybe it would make sense to resolve https://github.com/falconry/falcon/issues/1514 first and then come back to this PR? I looked into how werkzeug does it, and it’s super-simple, so I think I could change it pretty quickly. Catching all of the references to LIFO order in the docs seems like the hardest part :joy:. That issue is still in “Needs Decision” status, but I think it would be a significant usability win (assuming there are no objections to storing the error handlers in a dict).

File: falcon/api.py (hunk 1 above)
Comment (jmvrbanac) [reply falcon/api.py]: Yeah, we're good with `Exception`, my bad.

**Comment by kgriffs** [comment]
Perhaps we should disable the hug test temporarily and open an issue over there? We can re-enable the hug tests in 3.1 once all of Falcon's breaking changes have been addressed. @timothycrosley what would you suggest?

**Comment by timothycrosley** [comment]
@kgriffs I think that sounds good! Once Falcon approaches a release candidate for the next version I'll start working on resolving any breaking issues that arise. Once I feel like hug is in good shape to support the new version I can make a pull request to uncomment the tests.

**Comment by vytas7** [comment]
Thanks for input @timothycrosley !
And :+1: from me for this path, we are probably going to hit more breaking changes on the road to 3.0 anyway.

Bandwidth permitting, I will try to submit a Hug PR if the changes are trivial (as I did for Falcon 2.0), otherwise we are going to need your help @timothycrosley :slightly_smiling_face:

**Comment by kgriffs** [comment]
OK, sounds like a plan. I will disable the hug test for the moment.

**Comment by kgriffs** [comment]
Depends on #1599

File: docs/user/faq.rst [hunk 3]
```diff
@@ -446,9 +446,14 @@ for this exception type. Alternatively, you may be able to configure your web
 server to transform the response for you (e.g., using Nginx's ``error_page``
 directive).
 
-500 errors are typically the result of an unhandled exception making its way
-up to the web server. To handle these errors more gracefully, you can add a
-custom error handler for Python's base :class:`Exception` type.
+By default, Falcon handles all exceptions not inheriting from
```
Comment (vytas7) [inline docs/user/faq.rst]: Technically, _all exceptions_ is not 100% correct since we are not installing any handler for `BaseException` (neither should we). But probably we don't need to dig that deep in the FAQ, this will do :+1:

Comment (csojinb) [reply docs/user/faq.rst]: Fair point. I have to fix the commit message style anyway, so I can go make this technically correct

Comment (vytas7) [reply docs/user/faq.rst]: Oh, sorry for stepping on your toes... I have hopefully addressed Sphinx complaints via the Web UI

Comment (csojinb) [reply docs/user/faq.rst]: No worries :-)

Comment (vytas7) [reply docs/user/faq.rst]: And btw don't worry about the commit message style. We maintainers get an extra chance to edit that when squash-merging a pull request. (But we seldom get it right ourselves...)

Comment (csojinb) [reply docs/user/faq.rst]: Okay, I'm just double-checking that the docs are formatted how I wanted, and then I'll push and we should be good to go

**Comment by csojinb** [comment]
Okay @vytas7, I modified the documentation to be more technically correct re: default error-handling. If it all looks good to you, then I'm happy with it.
